## Imports

In [1]:
import pandas as pd

## Data Loading

In [2]:
df_features= pd.read_csv('data/train_features.csv')
df_labels= pd.read_csv('data/train_labels.csv')

## EDA And Data Cleaning

Comparing the length of the target and the feature dataframes, if they are the same length as expected, The 2 data frames shall be concatenated so as to clean the data effectively.

In [3]:
len(df_features) == len(df_labels)

True

In [4]:
# concatenating the dataframes
df = pd.concat([df_features, df_labels], axis=1)
df.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  ... quality_group  \
0  34.938093  -9.856322                  none            0  ...          good   
1  34.698766  -2.147466              Zahanati            0  ...          good   
2  37.460664  -3.821329           Kwa Mahundi            0  ...          good   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0  ...          good   
4  31.130847  -1.825359               Shuleni            0  ...          good   

       quantity quantity_group                source           source_type  \
0        enough         enough                spring                spring   
1  insufficient   insufficient  rainwater harvesting  rainwater harvesting   
2        enough         enough                   dam                   dam   
3           dry            dry           machine dbh              borehole   
4      seasonal       seasonal  rainwater harvesting  rainwater harvesting   

  source_class              waterpoint_type  waterpoint_type_group     id  \
0  groundwater           communal standpipe     communal standpipe  69572   
1      surface           communal standpipe     communal standpipe   8776   
2      surface  communal standpipe multiple     communal standpipe  34310   
3  groundwater  communal standpipe multiple     communal standpipe  67743   
4      surface           communal standpipe     communal standpipe  19728   

     status_group  
0      functional  
1      functional  
2      functional  
3  non functional  
4      functional  

[5 rows x 42 columns]

In [5]:
#inspecting the target variables
df['status_group'].value_counts()

functional                 32259
non functional             22824
functional needs repair     4317
Name: status_group, dtype: int64

This is ternary classification problem since the target has 3 possible values

## Exploring the data (EDA)


df_features.columns

In [6]:
df_features.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga',
       'ward', 'population', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'construction_year',
       'extraction_type', 'extraction_type_group', 'extraction_type_class',
       'management', 'management_group', 'payment', 'payment_type',
       'water_quality', 'quality_group', 'quantity', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group'],
      dtype='object')

### The Features in This Dataset

The following set of information about the waterpoints is provided:

- **amount_tsh**: Total static head (amount of water available to the waterpoint)
- **date_recorded**: The date the row was entered
- **funder**: Who funded the well
- **gps_height**: Altitude of the well
- **installer**: Organization that installed the well
- **longitude**: GPS coordinate
- **latitude**: GPS coordinate
- **wpt_name**: Name of the waterpoint, if there is one
- **num_private**: _(Not specified)_
- **basin**: Geographic water basin
- **subvillage**: Geographic location
- **region**: Geographic location
- **region_code**: Geographic location (coded)
- **district_code**: Geographic location (coded)
- **lga**: Geographic location
- **ward**: Geographic location
- **population**: Population around the well
- **public_meeting**: True/False, indicating if a public meeting was held
- **recorded_by**: Group entering this row of data
- **scheme_management**: Who operates the waterpoint
- **scheme_name**: The name of the management scheme
- **permit**: Whether the waterpoint is permitted (True/False)
- **construction_year**: Year the waterpoint was constructed
- **extraction_type**: The kind of extraction method the waterpoint uses
- **extraction_type_group**: The extraction type grouped by similar methods
- **extraction_type_class**: The extraction type categorized into broader classes
- **management**: How the waterpoint is managed
- **management_group**: The management grouped into broader categories
- **payment**: The cost structure of the water
- **payment_type**: The specific type of payment method
- **water_quality**: The quality of the water
- **quality_group**: The quality grouped by similar qualities
- **quantity**: The quantity of water
- **quantity_group**: The quantity grouped by similar quantities
- **source**: The source of the water
- **source_type**: The type of water source
- **source_class**: The broader class of the water source
- **waterpoint_type**: The specific type of waterpoint
- **waterpoint_type_group**: The type of waterpoint grouped by similar types


In [7]:
df.sample(n=10)

id  amount_tsh date_recorded                  funder  gps_height  \
30394  62434         0.0    2011-08-05                   Hifab           0   
4556   14991         0.0    2011-04-08                     NaN           0   
27721  60497         0.0    2011-07-06                  Hewasa           0   
22604  12023         0.0    2012-10-28                   Tahea           0   
55525  17196         0.0    2011-08-22                   Tasaf        1739   
50041   3375         0.0    2011-07-13                      Po        1551   
15160  73068         0.0    2011-07-27  Government Of Tanzania        1255   
36453  58273         0.0    2011-07-27  Government Of Tanzania           0   
6440   37180         0.0    2013-02-02                    Finw         377   
10395  19385         0.0    2011-07-27                      Il        1666   

                installer  longitude   latitude                    wpt_name  \
30394              Hesawa  33.482611  -2.784059                  Mwabakanga   
4556                  NaN  34.369072  -8.698171               Kwa Mzee Musa   
27721                 DWE  31.517362  -1.601114     Shule Ya Msingi Nyakaju   
22604               TAHEA  32.132423  -5.021029                     Manyala   
55525                 DWE  31.706724  -8.595941  Shamba La Bernard Kapikita   
50041                PRIV  31.030281  -7.493705                      Polisi   
15160                 RWE  33.018703  -2.079060                        none   
36453  central government  33.864142  -9.430502           Kwa Mose Mwamsojo   
6440                 FinW  39.448854 -10.622765                    Zahanati   
10395                COUN  31.265001  -7.772213                      Mayeka   

       num_private  ... quality_group      quantity quantity_group  \
30394            0  ...          good        enough         enough   
4556             0  ...       colored      seasonal       seasonal   
27721            0  ...          good        enough         enough   
22604            0  ...          good        enough         enough   
55525            0  ...          good        enough         enough   
50041            0  ...         salty        enough         enough   
15160            0  ...          good  insufficient   insufficient   
36453            0  ...          good           dry            dry   
6440             0  ...          good        enough         enough   
10395            0  ...          good  insufficient   insufficient   

             source   source_type source_class              waterpoint_type  \
30394  shallow well  shallow well  groundwater                    hand pump   
4556          river    river/lake      surface           communal standpipe   
27721        spring        spring  groundwater           communal standpipe   
22604  shallow well  shallow well  groundwater                    hand pump   
55525   machine dbh      borehole  groundwater                    hand pump   
50041   machine dbh      borehole  groundwater                    hand pump   
15160          lake    river/lake      surface  communal standpipe multiple   
36453        spring        spring  groundwater           communal standpipe   
6440    machine dbh      borehole  groundwater           communal standpipe   
10395         river    river/lake      surface           communal standpipe   

       waterpoint_type_group     id             status_group  
30394              hand pump  62434               functional  
4556      communal standpipe  14991  functional needs repair  
27721     communal standpipe  60497  functional needs repair  
22604              hand pump  12023           non functional  
55525              hand pump  17196               functional  
50041              hand pump   3375               functional  
15160     communal standpipe  73068           non functional  
36453     communal standpipe  58273           non functional  
6440      communal standpipe  37180  functional needs repair  
10395     

- On inspection, some columns e.g `quantity` and `quantity_group` seem to have identical information. Thus, we first verify if indeed this is the case

In [8]:
# Check if columns are identical
are_identical = (df['quantity'] == df['quantity_group']).all()

print("Columns are identical:", are_identical)

Columns are identical: True


### Column Inspection Comparison

To streamline our dataset and remove redundancy, we will perform the following steps:

1. **Identify Columns with Similar Names or Descriptions**  
   We'll start by identifying columns that have similar names or descriptions. This helps us determine which columns might contain overlapping information.

2. **Compare Column Content**  
   For columns identified as similar, we will conduct further checks to compare their content. This involves:
   - **Exact Equality Check**: Verifying if the values in the columns are identical for each row.
   - **Handling Missing Values**: Ensuring that missing values are considered in the comparison.
   
3. **Trim Down the Dataset**  
   Based on the comparison results, we will:
   - **Remove Redundant Columns**: Eliminate columns that contain duplicate or redundant information.
   - **Optimize the Dataset**: Reduce the dataset size and complexity by retaining only the necessary columns.

By following these steps, we will improve the quality and efficiency of our dataset, ensuring that we work with clean and relevant data.


### The Features in This Dataset

The following set of information about the waterpoints is provided:

- **amount_tsh**: Total static head (amount of water available to the waterpoint)
- **date_recorded**: The date the row was entered
- **funder**: Who funded the well
- **gps_height**: Altitude of the well
- **installer**: Organization that installed the well
- **longitude**: GPS coordinate
- **latitude**: GPS coordinate
- **wpt_name**: Name of the waterpoint, if there is one
- **num_private**: _(Not specified)_
- **basin**: Geographic water basin
- **subvillage**: Geographic location
- **region**: Geographic location
- **region_code**: Geographic location (coded)
- **district_code**: Geographic location (coded)
- **lga**: Geographic location, Local Government Authorities (LGAs) in Tanzania are accountable for the delivery of public services to citizens - providing oversight and management support to health facilities, schools, and villages.
- **ward**: Geographic location
- **population**: Population around the well
- **public_meeting**: True/False, indicating if a public meeting was held
- **recorded_by**: Group entering this row of data
- **scheme_management**: Who operates the waterpoint
- **scheme_name**: The name of the management scheme
- **permit**: Whether the waterpoint is permitted (True/False)
- **construction_year**: Year the waterpoint was constructed
- **extraction_type**: The kind of extraction method the waterpoint uses
- **extraction_type_group**: The extraction type grouped by similar methods
- **extraction_type_class**: The extraction type categorized into broader classes
- **management**: How the waterpoint is managed
- **management_group**: The management grouped into broader categories
- **payment**: The cost structure of the water
- **payment_type**: The specific type of payment method
- **water_quality**: The quality of the water
- **quality_group**: The quality grouped by similar qualities
- **quantity**: The quantity of water
- **quantity_group**: The quantity grouped by similar quantities
- **source**: The source of the water
- **source_type**: The type of water source
- **source_class**: The broader class of the water source
- **waterpoint_type**: The specific type of waterpoint
- **waterpoint_type_group**: The type of waterpoint grouped by similar types


In [9]:
df.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga',
       'ward', 'population', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'construction_year',
       'extraction_type', 'extraction_type_group', 'extraction_type_class',
       'management', 'management_group', 'payment', 'payment_type',
       'water_quality', 'quality_group', 'quantity', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group', 'id', 'status_group'],
      dtype='object')

In [10]:
locations = ['subvillage','region','region_code','district_code','lga','ward']
df[locations].head()

subvillage   region  region_code  district_code        lga        ward
0    Mnyusi B   Iringa           11              5     Ludewa    Mundindi
1     Nyamara     Mara           20              2  Serengeti       Natta
2     Majengo  Manyara           21              4  Simanjiro     Ngorika
3  Mahakamani   Mtwara           90             63   Nanyumbu    Nanyumbu
4  Kyanyamisa   Kagera           18              1    Karagwe  Nyakasimbi

In [11]:
df['region_code'].describe()

count    59400.000000
mean        15.297003
std         17.587406
min          1.000000
25%          5.000000
50%         12.000000
75%         17.000000
max         99.000000
Name: region_code, dtype: float64

In [12]:
df['district_code'].describe()

count    59400.000000
mean         5.629747
std          9.633649
min          0.000000
25%          2.000000
50%          3.000000
75%          5.000000
max         80.000000
Name: district_code, dtype: float64

In [13]:
for location in locations:
    print(f"Column : {location}, Unique values : {len(df[location].unique())}")


Column : subvillage, Unique values : 19288
Column : region, Unique values : 21
Column : region_code, Unique values : 27
Column : district_code, Unique values : 20
Column : lga, Unique values : 125
Column : ward, Unique values : 2092


In [14]:
df[locations].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   subvillage     59029 non-null  object
 1   region         59400 non-null  object
 2   region_code    59400 non-null  int64 
 3   district_code  59400 non-null  int64 
 4   lga            59400 non-null  object
 5   ward           59400 non-null  object
dtypes: int64(2), object(4)
memory usage: 2.7+ MB


- All this seem to show information about geographical location. From inspection, `district_code` and `region_code` seem to be the most suitable since they are codes of locations and have relatively fewer classes (80 for `district_code`, 100 for `region_code`)

In [15]:
#region and region code are very similar let's explore further..


In [16]:
unique_regions = df['region'].unique()
for region in unique_regions:
    print(region)

Iringa
Mara
Manyara
Mtwara
Kagera
Tanga
Shinyanga
Tabora
Pwani
Ruvuma
Kilimanjaro
Rukwa
Mwanza
Kigoma
Lindi
Dodoma
Arusha
Mbeya
Singida
Morogoro
Dar es Salaam


In [17]:
unique_region_codes = df['region_code'].unique()
for region in unique_region_codes:
    print(region)

11
20
21
90
18
4
17
14
60
10
3
15
19
16
80
1
6
2
12
13
5
7
99
24
9
8
40


## Column Selection for Logistic Regression and Decision Tree Models

### 1. Categorical Variables

- **Decision Trees**: Handle categorical variables well. You can use all categorical columns directly as decision trees can split on different categories without numerical encoding.
  
- **Logistic Regression**: Requires numerical input. Categorical variables need to be encoded into numerical formats, typically using one-hot encoding or label encoding.

### 2. Column Selection

- **`subvillage`**: 19,288 unique values
  - **Note**: High cardinality may make this column challenging to encode and could lead to overfitting. Consider not using it directly.

- **`region`**: 21 unique values
  - **Note**: Moderate number of categories. Can be encoded using one-hot encoding or label encoding.

- **`region_code`**: 27 unique values
  - **Note**: Similar to `region`, this column has a manageable number of categories and can be used after encoding.

- **`district_code`**: 20 unique values
  - **Note**: Manageable number of categories. Suitable for encoding.

- **`lga`**: 125 unique values
  - **Note**: High number of unique values, which may be challenging for encoding and could lead to overfitting.

- **`ward`**: 2,092 unique values
  - **Note**: Very high cardinality, potentially problematic for encoding and may not provide much predictive value.

### Recommendations

- **Decision Trees**: You can use all columns directly. Decision trees handle categorical variables well and can manage high-cardinality features without extensive preprocessing.

- **Logistic Regression**:
  - **Use Columns with Fewer Unique Values**: Start with `region`, `region_code`, `district_code`, and `lga` due to their more manageable number of unique values.
  - **Consider Feature Reduction**: Avoid using `subvillage` and `ward` directly because of their high cardinality. Consider aggregating or grouping these columns to reduce the number of unique values.
  - **One-Hot Encoding**: Apply one-hot encoding for categorical columns that have a manageable number of unique values.


For analysis we shall use  `region_code`, `district_code`, and `lga` due to the reasons stated above and trim `region`,`subvillage` and `ward` columns


In [18]:
redundant_cols = ['region','subvillage','ward']
df_concise = df.drop(redundant_cols, axis = 1)


In [19]:
df_concise.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'region_code', 'district_code', 'lga', 'population',
       'public_meeting', 'recorded_by', 'scheme_management', 'scheme_name',
       'permit', 'construction_year', 'extraction_type',
       'extraction_type_group', 'extraction_type_class', 'management',
       'management_group', 'payment', 'payment_type', 'water_quality',
       'quality_group', 'quantity', 'quantity_group', 'source', 'source_type',
       'source_class', 'waterpoint_type', 'waterpoint_type_group', 'id',
       'status_group'],
      dtype='object')

### The Features in This Dataset

The following set of information about the waterpoints is provided:

- **amount_tsh**: Total static head (amount of water available to the waterpoint)
- **date_recorded**: The date the row was entered
- **funder**: Who funded the well
- **gps_height**: Altitude of the well
- **installer**: Organization that installed the well
- **longitude**: GPS coordinate
- **latitude**: GPS coordinate
- **wpt_name**: Name of the waterpoint, if there is one
- **num_private**: _(Not specified)_
- **basin**: Geographic water basin
- **subvillage**: Geographic location
- **region**: Geographic location
- **region_code**: Geographic location (coded)
- **district_code**: Geographic location (coded)
- **lga**: Geographic location, Local Government Authorities (LGAs) in Tanzania are accountable for the delivery of public services to citizens - providing oversight and management support to health facilities, schools, and villages.
- **ward**: Geographic location
- **population**: Population around the well
- **public_meeting**: True/False, indicating if a public meeting was held
- **recorded_by**: Group entering this row of data
- **scheme_management**: Who operates the waterpoint
- **scheme_name**: The name of the management scheme
- **permit**: Whether the waterpoint is permitted (True/False)
- **construction_year**: Year the waterpoint was constructed
- **extraction_type**: The kind of extraction method the waterpoint uses
- **extraction_type_group**: The extraction type grouped by similar methods
- **extraction_type_class**: The extraction type categorized into broader classes
- **management**: How the waterpoint is managed
- **management_group**: The management grouped into broader categories

- **payment**: The cost structure of the water
- **payment_type**: The specific type of payment method

- **water_quality**: The quality of the water
- **quality_group**: The quality grouped by similar qualities

- **quantity**: The quantity of water
- **quantity_group**: The quantity grouped by similar quantities

- **source**: The source of the water
- **source_type**: The type of water source
- **source_class**: The broader class of the water source

- **waterpoint_type**: The specific type of waterpoint
- **waterpoint_type_group**: The type of waterpoint grouped by similar types


Next up...
- **extraction_type**: The kind of extraction method the waterpoint uses
- **extraction_type_group**: The extraction type grouped by similar methods
- **extraction_type_class**: The extraction type categorized into broader classes

In [20]:
extraction = ['extraction_type','extraction_type_group','extraction_type_class']
df_concise[extraction].head()

extraction_type extraction_type_group extraction_type_class
0         gravity               gravity               gravity
1         gravity               gravity               gravity
2         gravity               gravity               gravity
3     submersible           submersible           submersible
4         gravity               gravity               gravity

-They seem to be exact, let's test whether they are identical.

In [21]:
# Check if columns are identical
are_identical = (df_concise['extraction_type_class'] == df_concise['extraction_type_group']).all()

print("Columns are identical:", are_identical)

Columns are identical: False


In [22]:
for element in extraction:
    print(f"Column : {element}, Unique values : {len(df[element].unique())}")

Column : extraction_type, Unique values : 18
Column : extraction_type_group, Unique values : 13
Column : extraction_type_class, Unique values : 7


In [23]:
unique_extraction_types = df['extraction_type'].unique()
for element in unique_extraction_types:
    print(element)

gravity
submersible
swn 80
nira/tanira
india mark ii
other
ksb
mono
windmill
afridev
other - rope pump
india mark iii
other - swn 81
other - play pump
cemo
climax
walimi
other - mkulima/shinyanga


In [24]:
unique_extraction_types = df['extraction_type_group'].unique()
for element in unique_extraction_types:
    print(element)

gravity
submersible
swn 80
nira/tanira
india mark ii
other
mono
wind-powered
afridev
rope pump
india mark iii
other handpump
other motorpump


In [25]:
unique_extraction_types = df['extraction_type_class'].unique()
for element in unique_extraction_types:
    print(element)

gravity
submersible
handpump
other
motorpump
wind-powered
rope pump


# IMP
- we settle on `extraction_type_class`


In [26]:
df_concise = df_concise.drop(['extraction_type','extraction_type_group'], axis = 1)

Next up...
- **management**: How the waterpoint is managed
- **management_group**: The management grouped into broader categories

In [27]:
management = ['management','management_group']
df_concise[management].head()

management management_group
0        vwc       user-group
1        wug       user-group
2        vwc       user-group
3        vwc       user-group
4      other            other

In [28]:
for element in management:
    print(f"Column : {element}, Unique values : {len(df[element].unique())}")

Column : management, Unique values : 12
Column : management_group, Unique values : 5


In [29]:
unique_management = df['management'].unique()
for element in unique_management:
    print(element)

vwc
wug
other
private operator
water board
wua
company
water authority
parastatal
unknown
other - school
trust


In [30]:
unique_management = df['management_group'].unique()
for element in unique_management:
    print(element)

user-group
other
commercial
parastatal
unknown


# IMP
- we settle on `management_group`


In [31]:
df_concise = df_concise.drop(['management'], axis = 1)
df_concise.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'region_code', 'district_code', 'lga', 'population',
       'public_meeting', 'recorded_by', 'scheme_management', 'scheme_name',
       'permit', 'construction_year', 'extraction_type_class',
       'management_group', 'payment', 'payment_type', 'water_quality',
       'quality_group', 'quantity', 'quantity_group', 'source', 'source_type',
       'source_class', 'waterpoint_type', 'waterpoint_type_group', 'id',
       'status_group'],
      dtype='object')

Next up...
- **payment**: The cost structure of the water
- **payment_type**: The specific type of payment method

In [37]:
payment = ['payment', 'payment_type']

In [41]:
def column_info(similar_cols):
    for element in similar_cols:
        print(f"Column : {element}, Unique values : {len(df[element].unique())}")
    for element in range(len(similar_cols)):
        print (f"Column: {similar_cols[element]}")
        for row in df_concise[similar_cols[element]].unique():
            print(row)
        print('\n')
column_info(payment)
    

Column : payment, Unique values : 7
Column : payment_type, Unique values : 7
Column: payment
pay annually
never pay
pay per bucket
unknown
pay when scheme fails
other
pay monthly


Column: payment_type
annually
never pay
per bucket
unknown
on failure
other
monthly




# IMP
- we settle on `payment_type`


In [47]:
df_concise = df_concise.drop(['payment'], axis = 1)


KeyError: "['payment'] not found in axis"

In [48]:
df_concise.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'region_code', 'district_code', 'lga', 'population',
       'public_meeting', 'recorded_by', 'scheme_management', 'scheme_name',
       'permit', 'construction_year', 'extraction_type_class',
       'management_group', 'payment_type', 'water_quality', 'quality_group',
       'quantity', 'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group', 'id', 'status_group'],
      dtype='object')

Next up...
- **water_quality**: The quality of the water
- **quality_group**: The quality grouped by similar qualities


In [42]:
water_quality = ['water_quality','quality_group']
column_info(water_quality)

Column : water_quality, Unique values : 8
Column : quality_group, Unique values : 6
Column: water_quality
soft
salty
milky
unknown
fluoride
coloured
salty abandoned
fluoride abandoned


Column: quality_group
good
salty
milky
unknown
fluoride
colored




# IMP
- we settle on `quality_group`


In [51]:
df_concise = df_concise.drop(['water_quality'], axis = 1)
df_concise.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'region_code', 'district_code', 'lga', 'population',
       'public_meeting', 'recorded_by', 'scheme_management', 'scheme_name',
       'permit', 'construction_year', 'extraction_type_class',
       'management_group', 'payment_type', 'quality_group', 'quantity',
       'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group', 'id', 'status_group'],
      dtype='object')

Next up...
- **quantity**: The quantity of water
- **quantity_group**: The quantity grouped by similar quantities

In [43]:
water_quanity = ['quantity','quantity_group']
column_info(water_quanity)

Column : quantity, Unique values : 5
Column : quantity_group, Unique values : 5
Column: quantity
enough
insufficient
dry
seasonal
unknown


Column: quantity_group
enough
insufficient
dry
seasonal
unknown




In [44]:
# Check if columns are identical
are_identical = (df_concise['quantity'] == df_concise['quantity_group']).all()

print("Columns are identical:", are_identical)

Columns are identical: True


# IMP
- we settle on `quantity_group`


In [52]:
df_concise = df_concise.drop(['quantity'], axis = 1)
df_concise.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'region_code', 'district_code', 'lga', 'population',
       'public_meeting', 'recorded_by', 'scheme_management', 'scheme_name',
       'permit', 'construction_year', 'extraction_type_class',
       'management_group', 'payment_type', 'quality_group', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group', 'id', 'status_group'],
      dtype='object')

Next up...
- **source**: The source of the water
- **source_type**: The type of water source
- **source_class**: The broader class of the water source

In [45]:
source = ['source','source_type','source_class']
column_info(source)

Column : source, Unique values : 10
Column : source_type, Unique values : 7
Column : source_class, Unique values : 3
Column: source
spring
rainwater harvesting
dam
machine dbh
other
shallow well
river
hand dtw
lake
unknown


Column: source_type
spring
rainwater harvesting
dam
borehole
other
shallow well
river/lake


Column: source_class
groundwater
surface
unknown


